path_login.go

```diff
@@ -39,15 +39,68 @@ has included a signature.`,
 			},
 		},
 		Callbacks: map[logical.Operation]framework.OperationFunc{
-			logical.UpdateOperation: b.pathLoginUpdate,
+			logical.UpdateOperation:      b.pathLoginUpdate,
+			logical.ResolveRoleOperation: b.pathLoginResolveRole,
 		},
 		HelpSynopsis:    pathLoginSyn,
 		HelpDescription: pathLoginDesc,
 	}
 }
 
-func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+func (b *backend) pathLoginResolveRole(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
```

Comment:
Maybe worth just adding a comment explaining that there is lots of duplicated logic here with the `pathLoginUpdate` function, and it's so that this newer endpoint can return logical errors when appropriate instead of internal server errors.

Reply:
Sounds good - have done so :)